feat: Add Wear OS Interface for GPSTest

adaext opened the pull request:
A Wear OS Interface Draft.

Closes https://github.com/barbeau/gpstest/issues/560

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/barbeau/gpstest?pullRequest=605) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/barbeau/gpstest?pullRequest=605) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/barbeau/gpstest?pullRequest=605) it.</sub>

barbeau commented on gpswearos/build.gradle:
```diff
@@ -0,0 +1,68 @@
+plugins {
```
Could you please change the werable folder name to just `wear` instead of `gpswearos`?

adaext replied:
> Could you please change the wearable folder name to just `wear` instead of `gpswearos`?

Ok, thanks!

barbeau commented on gpswearos/build.gradle:
```diff
@@ -0,0 +1,68 @@
+
+    defaultConfig {
+        applicationId "com.android.gpswearos"
+        minSdk 30
```
What is the lowest minSdk you can use with Wear? The app has a minSdk of 24, so ideally we support back to that, or whatever the minimum for Wear is.

barbeau replied:
Looks like lowest minSdk for Wear 2.0 is 25, so let's use that - https://developer.android.com/training/wearables/upgrade#version

adaext replied:
> Looks like lowest minSdk for Wear 2.0 is 25, so let's use that - https://developer.android.com/training/wearables/upgrade#version

Ok, I will set the minSdk with 25.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/SampleData.kt:
```diff
@@ -0,0 +1,15 @@
+object SampleData {
+    // Sample conversation data
+    val statusSample = listOf(
+        SatelliteStatus("Lat:", "28.0713003 °"),
```
These fields will all be derived for a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationRow of something similar.

Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.

adaext replied:
Ok, thanks

adaext replied:
> These fields will all be derived from a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationLabelAndData or something similar.
> 
> Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.

Ok, thanks!

adaext replied:
> > These fields will all be derived from a `Location` object. SatelliteStatus will hold the info from the GnssStatus object. So I would change this container class to LocationLabelAndData or something similar.
> > Also, you should be able to move some of this common code and resources into a shared module. SatelliteStatus could go there, as well as the strings for lat, lon, etc.
> 
> Ok, thanks!

I've created a library module to put sharing data. I've added the module dependency successfully to `wear` module but failed to add it to GPSTest. It says "No build type in module ':library' matches build type 'android'.". Since I haven't figured out how to solve it, I will just apply this to `wear` in the next commit.

barbeau replied:
Ok, thanks. Could you work on this a bit more to see if you can get it to work for the app too? Otherwise there is going to be a lot of code duplication between wear and the app once we get into listening for GNSS updates.

See this for guidance - https://developer.android.com/studio/projects/android-library

barbeau replied:
Once you get the library set up, you can move all the classes in the normal app `.di`, `.data`, `.model`, and `.util` packages to the library (keep them in the respective subpackages), as those will be used to get live GNSS data on Wear (and in the app).

So you should end up having in the library:
* `com.android.gpstest.library.di`
* `com.android.gpstest.library.data`
* `com.android.gpstest.library.model`
* `com.android.gpstest.library.util`

adaext replied:
> So I worked on this a bit, and it turns out the app just needs to be updated in a few places due to other changes. Here's a working branch showing the data structure in the library being rendered in the normal app Status screen under the LocationCard: https://github.com/barbeau/gpstest/tree/wear-working-library
> 
> Here's the commit with the changes I made to get the library to work with the app: [3ec9aff](https://github.com/barbeau/gpstest/commit/3ec9aff1dda1d5e781108ebf91ab25bd359e1349)
> 
> You can ignore the changes made to the wear MainActivity - those were part of another test I was doing and I accidentally included them in the commit.

Thanks! It works now!

adaext replied:
> Once you get the library set up, you can move all the classes in the normal app `.di`, `.data`, `.model`, and `.util` packages to the library (keep them in the respective subpackages), as those will be used to get live GNSS data on Wear (and in the app).
> 
> So you should end up having in the library:
> 
> * `com.android.gpstest.library.di`
> * `com.android.gpstest.library.data`
> * `com.android.gpstest.library.model`
> * `com.android.gpstest.library.util`

I've followed this instruction and found some problems. I've moved the `.di`, `.data`, `.model`, and `.util` packages to the `library` at first, and found these use a lot of things in `res/values`. After copying the `res/values` to the current `library`, I found I have to move `Application` to `library` for the same reason. After this,  there are still some classes needed to move to `library`, After moving, I found more classes need to be moved since the new moved classes cite the classes in the origin package. Since I haven't figured out how to solve this, I will first commit the revisions under resolved conversations.

barbeau replied:
>I found I have to move Application to library for the same reason

Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.

Here is the approach I would recommend:
1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.

Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods.

If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

adaext replied:
> > I found I have to move Application to library for the same reason
> 
> Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.
> 
> Here is the approach I would recommend:
> 
> 1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
> 2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.
> 
> Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods. So, with this approach you shouldn't need the Application class in the library.
> 
> If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

Should the function `sendEmail()` inside `util.UIUtils.kt` be also moved to the library? `SignalInfoViewModel` was cited as a parameter of `sendEmail()`. If 'sendEmail()' was reserved, `SignalInfoViewModel` should also be moved to the library.

adaext replied:
> > I found I have to move Application to library for the same reason
> 
> Ah, yes, good point. So IIRC the Application class is mainly used throughout the app as a way to access the application-wide context to get strings from resources, etc. (as opposed to using the Activity context) and to access a centralized preferences object.
> 
> Here is the approach I would recommend:
> 
> 1. For all the classes in the `.data` package (*Manager.kt), I would replace the reference to `Application.app` with the `context` that is passed into the constructor
> 2. For all the classes in the `.util` package that use `Application.Companion.getPrefs()` (in Java classes) or `Application.prefs` (in Kotlin classes), I would refactor the methods that use them to take in a `SharedPreferences` as a new parameter.
> 
> Then, the wear application and mobile application can both implement their own version of the `Application` class (which will be very similar), and pass in the `Application.app` for the context and `Application.prefs` for the SharedPreferences to all the relevant methods. So, with this approach you shouldn't need the Application class in the library.
> 
> If you follow this approach, are there any other classes in GPSTest module that are still needed in the library?

Should the function `sendEmail()` inside `util.UIUtils.kt` be also moved to the library? `SignalInfoViewModel` was referenced as a parameter of `sendEmail()`. If 'sendEmail()' was reserved, `SignalInfoViewModel` should also be moved to the library.

barbeau replied:
I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

adaext replied:
> 

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

adaext replied:
> `ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

adaext replied:
> I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

`ShareDialogFragment` in `gpstest.ui.share` was used in `UIUtils.kt`. Should code about Fragment be moved out of `UIUtils.kt` and put back to `gpstest`?

adaext replied:
> I believe `SignalInfoViewModel` should work on Wear, so I'd say yes, let's go ahead and move it over. It will help processing data from the GNSS listeners and preparing it for the UI (it powers the Status view in the normal app).

`GnssFilterDialog` and `HelpActivity` also seem shouldn't be shared in `library`. Should I move these to `gpstest`? And should `res` be copied to `library`? The content in `res` has been referenced many times in the newly moved library content.

barbeau replied:
Yes, Dialogs, Fragments, and Activities should stay in the app. For the util methods that use these, they should stay in the utility method class in the app.

barbeau replied:
And yes, at least do_not_translate.xml, and likely some other strings, and possibly some other res (flag images) should move to library.

adaext replied:
> And yes, at least do_not_translate.xml, and likely some other strings, and possibly some other res (flag images) should move to library.

Thanks. Now, some resources are in GPSTest, Others are in library. However, some codes in library can't find the resources even if the resources are in library too. At first, the compiler couldn't recognize the resources in library and marked this with red line. I've added `xmlns:tools="http://schemas.android.com/tools" tools:ignore="ExtraTranslation"` inside resource tag, and the red line disappeared. The current problem is that the compiler could find the resources but the Gradle couldn't. When I try to build the project, the build results show "Unresolved reference ..."

barbeau replied:
You'll need to make sure that for each string from values/strings.xml and array from values/arrays.xml moved to the library, the corresponding strings are moved for all the translations in the `values-XX` (e.g., `values-es`) folders. In other words, each translation of a string must be in the same project as the English string. If you do this correctly you shouldn't need the `ExtraTranslation` tag.

>the compiler could find the resources but the Gradle couldn't

Could you explain more about this? By "compiler" do you mean Android Studio?

Please try to move all the strings in the right place, and if you're still having problems please push what you have to GitHub and I can take a look.

adaext replied:
> You'll need to make sure that for each string from values/strings.xml and array from values/arrays.xml moved to the library, the corresponding strings are moved for all the translations in the `values-XX` (e.g., `values-es`) folders. In other words, each translation of a string must be in the same project as the English string. If you do this correctly you shouldn't need the `ExtraTranslation` tag.
> 
> > the compiler could find the resources but the Gradle couldn't
> 
> Could you explain more about this? By "compiler" do you mean Android Studio?
> 
> Please try to move all the strings in the right place, and if you're still having problems please push what you have to GitHub and I can take a look.

Thanks a lot. I've selected the Android tab to view the whole project structure, which caused me to move these XML files to the wrong packages. Now I've solved this problem and selected the Project tab.

adaext replied:
As we discussed before, I added `SharedPreferences` to the constructors of a few classes. In order to make dependency-injection available, I've created another `@Provides`-annotated method in DataModule.kt to instantiate the required `SharedPreferences`. That method requires a `Context`, which used to be available to `DataModule` already. But the toolchain says that I'll need another `@Provides`-annotated method to create the required `Context`. In the repository before my change, there's no `@Provide`-annotated method that returns `Context` at all. The application context is just accessible directly.  I'm not sure if it's because I moved `DataModule` into library module so it lost such access. I've commit the new version and need your advice to do such cross-module dependency injections. Thanks!

barbeau replied:
@adaext Nice work! Look like you just need to add a `@Provides` method for the context itself in the `DataModule` method - like this:

```kotlin
    @Provides
    fun provideContext(
        @ApplicationContext context: Context,
    ): Context {
        return context
    }
```

I made that change and reduced `targetSdkVersion` to `29` in this commit and it seems to be working:
https://github.com/barbeau/gpstest/tree/wear-provide-context-example

Here's the commit with the changes - https://github.com/barbeau/gpstest/commit/ea8adbcf034b3419984ec514355aa1921f4ca4a4

adaext replied:
> 

Thanks, it finally works. I reduced the `targetSdkVersion` to 31 because of the error from Gradle, "Google Play requires that apps target API level 31 or higher.". And it also works.

barbeau replied:
You'll need to keep `targetSdkVersion 29` for the app to work on Android 13 devices (for now), otherwise it will crash on an error for the PendingIntent `Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE`.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,104 @@
+
+@Composable
+fun StatusCard(satStatues: List<SatelliteStatus>) {
+    Card({} , modifier = Modifier
```
There is an extra lambda with the braces here you can remove.

adaext replied:
> There is an extra lambda with the braces here you can remove.

If I remove this, the compiler will say "No value passed for parameter 'onClick'". It seems that it shouldn't be removed.

adaext replied:
> There is an extra lambda with the braces here you can remove.

If I remove this, the compiler will say "No value passed for parameter 'onClick'". It seems the extra lambda shouldn't be removed. Wear OS seems to be different from Phone on this.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,104 @@
+        Box ( modifier = Modifier
+            .fillMaxSize()
+            .verticalScroll(rememberScrollState()), contentAlignment = Alignment.Center){
+            Row {
```
If you use ScalingLazyColumn here and adapt the list of Location elements accordingly, it will handle scaling items on and off the screen as you scroll. See https://youtu.be/B7D3G6tC9n0?t=697.

adaext replied:
> If you use ScalingLazyColumn here and adapt the list of Location elements accordingly, it will handle scaling items on and off the screen as you scroll. See https://youtu.be/B7D3G6tC9n0?t=697.

Thanks, I've applied it to this, and it works!

adaext replied:
> If you use ScalingLazyColumn here and adapt the list of Location elements accordingly, it will handle scaling items on and off the screen as you scroll. See https://youtu.be/B7D3G6tC9n0?t=697.

Thanks, it works.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,104 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContent {
+            StatusCard(SampleData.statusSample)
```
We should use `Scaffold` here as the top level Composable for the basic infrastructure of the Wear app - see https://youtu.be/B7D3G6tC9n0?t=905. This also lets you include `timeText` at the top of the screen, and gives you a position indicator for scrolling a ScalingLazyColumn.

barbeau replied:
A good code lab where you can see a complete Wear Compose app with Theme and Scaffold is https://developer.android.com/codelabs/compose-for-wear-os#0. A corresponding code-along video is https://www.youtube.com/watch?v=7U3FwzuqgRE.

adaext replied:
> We should use `Scaffold` here as the top level Composable for the basic infrastructure of the Wear app - see https://youtu.be/B7D3G6tC9n0?t=905. This also lets you include `timeText` at the top of the screen, and gives you a position indicator for scrolling a ScalingLazyColumn.
> 
> We can skip the whole navigation configuration for now because for the minimum viable product (MVP) of this Wear app it will just be a single screen that simulates what you see in the "Status" tab in the phone app.

Thanks, I've learned this and applied it.

barbeau commented on build.gradle:
```diff
@@ -1,13 +1,17 @@
 buildscript {
+    ext {
+        compose_version = '1.2.0-alpha05'
+        wear_compose_version = '1.0.0-alpha18'
```
It looks like the latest Wear Compose stable release is `1.0.2`, and the latest alpha release is `1.1.0-alpha07`:
https://developer.android.com/jetpack/androidx/releases/wear-compose

Let's go with `1.1.0-alpha07` for now, and if we encounter problems we can fall back to the latest stable `1.0.2`.

adaext replied:
> 

Thanks, `1.0.2` is ok. `1.1.0-alpha07` has caused a lot of conflicts.

barbeau commented on build.gradle:
```diff
@@ -1,13 +1,17 @@
+        compose_version = '1.2.0-alpha05'
+        wear_compose_version = '1.0.0-alpha18'
+    }
     ext.kotlin_version = '1.5.31'
```
If you update to `ext.kotlin_version` to `'1.6.10'`, I believe that should get rid of the errors you're seeing and you can add Lint back in (i.e., remove the lines that disable Lint). I was able to create a basic Wear template here that runs without disabling Lint - you can compare against this branch and see if there are any other differences:
https://github.com/barbeau/gpstest/tree/wear-template

adaext replied:
> If you update to `ext.kotlin_version` to `'1.6.10'`, I believe that should get rid of the errors you're seeing and you can add Lint back in (i.e., remove the lines that disable Lint). I was able to create a basic Wear template here that runs without disabling Lint - you can compare against this branch and see if there are any other differences: https://github.com/barbeau/gpstest/tree/wear-template

Thanks, but If I use version '1.6.10', the project can't pass compiling. It says "This version (1.1.0-beta02) of the Compose Compiler requires Kotlin version 1.5.31 but you appear to be using Kotlin version 1.6.10 which is not known to be compatible.  Please fix your configuration (or `suppressKotlinVersionCompatibilityCheck` but don't say I didn't warn you!).". I've created the module by creating an empty compose Activity for Wear OS.  Have we created the new module in the same way?

barbeau replied:
Yes, I did the same thing, but  I had to manually bump the versions numbers myself and run the Gradle upgrade wizard. If you change all the version numbers I commented on together it should work. If you're still getting compilation errors after that you might need to invalidate caches and restart Android Studio.

adaext replied:
> bump

Thanks, I've updated the `kotlinCompilerExtensionVersion` to '1.1.0-rc02' and then updated the `ext.kotlin_version` to `1.6.10`. Now, it works well.

barbeau commented on gpswearos/build.gradle:
```diff
@@ -0,0 +1,68 @@
+
+    buildTypes {
+        release {
+            minifyEnabled false
```
`minifyEnabled` should be `true` for release builds.

adaext replied:
> `minifyEnabled` should be `true` for release builds.

Thanks, I will replace it.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,104 @@
+@Preview(device = Devices.WEAR_OS_SMALL_ROUND, showSystemUi = true)
+@Composable
+fun DefaultPreview() {
+    StatusCard(SampleData.statusSample)
```
Looks like this overflows on my preview screen and on the Wear OS Small round API 30 emulator:

![image](https://user-images.githubusercontent.com/928045/195119467-9c2ac298-b0be-4726-9b5b-484f37d4ce3e.png)

adaext replied:
> Looks like the text overflows each line on my preview screen and on the Wear OS Small round API 30 emulator:
> 
> ![wear-test](https://user-images.githubusercontent.com/928045/195119720-86ce16ac-a663-459c-87b6-462af027970e.png)
> 
> It will probably work better if we combine the label and value into a single string that's centered in the screen.

Ok, thanks. After doing this, I've found that the quotation marks can't be aligned vertically. 
![6MQ6B}8}L_PVU`)X25023~8](https://user-images.githubusercontent.com/69705892/196041650-fa2a4628-80b5-4f2f-8ebb-77f1312bd07f.png)

Does this seem ok?

barbeau replied:
>Ok, thanks. After doing this, I've found that the quotation marks can't be aligned vertically.

Yes, that looks fine to me. I think we need to make some compromises/design changes from the regular mobile version for the form factor of Wear, and this seems reasonable to me.

barbeau commented on gpswearos/src/main/java/com/android/gpswearos/presentation/SampleData.kt:
```diff
@@ -0,0 +1,15 @@
+    // Sample conversation data
+    val statusSample = listOf(
+        SatelliteStatus("Lat:", "28.0713003 °"),
+        SatelliteStatus("Time:", "04:35:29 PM"),
```
Let's see if we can remove time from the content list and instead populate the `timeText` in the Scaffold with the GPS time.

adaext replied:
> Let's see if we can remove time from the content list and instead populate the `timeText` in the Scaffold with the GPS time.

Ok, I've replaced it.

adaext commented on gpswearos/build.gradle:
```diff
@@ -0,0 +1,68 @@
+
+android {
+    namespace 'com.android.gpswearos'
+    compileSdk 32
```
> I believe compileSdk needs to be `33` with the latest Wear Compose library.

OK, thanks!

adaext commented on gpswearos/build.gradle:
```diff
@@ -0,0 +1,68 @@
+        compose true
+    }
+    composeOptions {
+        // kotlinCompilerExtensionVersion '1.1.1'
```
> This should be `1.1.1`

Ok, thanks.

adaext replied:
> This should be `1.1.1`

Thanks, but If I change this to `1.1.1`, I will get this compiling error message ":wear:compileDebugKotlin". Instead, version `1.1.0-beta02` works well.

barbeau commented on GPSTest/build.gradle:
```diff
@@ -6,6 +6,10 @@ apply plugin: 'dagger.hilt.android.plugin'
 android {
     compileSdkVersion 31
 
+    lint {
```
I believe you should be able to get rid of this now that you updated the dependencies?

adaext replied:
> I believe you should be able to get rid of this now that you updated the dependencies?

Sure, It works.

barbeau commented on library/src/main/java/com/android/library/LocationLabelAndData.kt:
```diff
@@ -0,0 +1,9 @@
+
+object LocationLabelAndData {
+    // Sample conversation data
+    val statusSample = listOf(
```
statusSample->locationLabelAndDataSample

adaext replied:
> statusSample->locationLabelAndDataSample

Ok, thanks.

barbeau commented on wear/src/main/java/com/android/wear/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,56 @@
+                        Text(text = satStatue)
+                    }
+                }
+            }
```
Inside the `ScalingLazyColumn` below the location info, let's add more `item {}`, one for each satellite in view from GnssStatus API:

<img width="178" alt="image" src="https://user-images.githubusercontent.com/928045/198314999-947c92f2-8cc0-4f14-967c-a726c2dca645.png">

Let's try to fit ID, GNSS, CF, C/N0, and Flags columns.

barbeau replied:
You should be able to copy/paste code from here for this section:
https://github.com/barbeau/gpstest/blob/master/GPSTest/src/main/java/com/android/gpstest/ui/status/StatusScreen.kt#L164

Or ideally, move it to the library module, although I saw your comment on having problems getting that working with the app.

adaext replied:
> You should be able to copy/paste code from here for this section: https://github.com/barbeau/gpstest/blob/master/GPSTest/src/main/java/com/android/gpstest/ui/status/StatusScreen.kt#L164
> 
> Or ideally, move it to the library module, although I saw your comment on having problems getting that working with the app.

Thanks, I'll do this after the code migrations from `GPSTest` to `library` work.

barbeau commented on wear/build.gradle:
```diff
@@ -0,0 +1,68 @@
+    compileSdk 33
+
+    defaultConfig {
+        applicationId "com.android.wear"
```
`com.android.gpstest.wear`

adaext replied:
> `com.android.gpstest.wear`

Ok, I will also set the namespace to the same value to coordinate.

barbeau commented on wear/src/main/java/com/android/wear/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,56 @@
+package com.android.wear.presentation
```
Can you please refactor this package/folder to `com.android.gpstest.wear`?

adaext replied:
Ok, thanks.

barbeau commented on library/build.gradle:
```diff
@@ -0,0 +1,41 @@
+}
+
+android {
+    namespace 'com.android.library'
```
Please change this and package to `com.android.gpstest.library`

adaext replied:
> Please change this and package/folder for classes to `com.android.gpstest.library`

Ok, thanks.

adaext commented on wear/src/main/java/com/android/wear/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,56 @@
+        Scaffold(
+            timeText = {
+                if (!listState.isScrollInProgress) {
+                    TimeText()
```
> Let's replace the default TimeText() with the GPS time formatted to the second.

Ok, thanks.

adaext commented on library/src/main/java/com/android/gpstest/library/theme/SignalInfoViewModel.kt:
```diff
@@ -13,28 +13,29 @@
  * limitations under the License.
  */
-package com.android.gpstest.ui
+package com.android.gpstest.library.theme
```
> Could you please move this class to a `library.ui` package instead of `library.theme`?

Sure, thanks!

barbeau commented:
In the latest version, looks like you were missing the Application name entry in the manifest. I added that and when launching you now see the dummy lat/lon values.

I also went ahead and added the Flows to the wear Activity. This should be very close to a working state in terms of listening to the GnssStatus and Location updates, although it needs some more debugging - I still don't see the status updates on the screen.

Here's a branch with the changes:
https://github.com/barbeau/gpstest/tree/wear-fix-application-manifest

...and the commit:
https://github.com/barbeau/gpstest/commit/3d440f234444a6a5c534674b4642a1a949799e1d

adaext commented:
> In the latest version, looks like you were missing the Application name entry in the manifest. I added that and when launching you now see the dummy lat/lon values.
> 
> I also went ahead and added the Flows to the wear Activity. This should be very close to a working state in terms of listening to the GnssStatus and Location updates, although it needs some more debugging - I still don't see the status updates on the screen.
> 
> Here's a branch with the changes: https://github.com/barbeau/gpstest/tree/wear-fix-application-manifest
> 
> ...and the commit: [3d440f2](https://github.com/barbeau/gpstest/commit/3d440f234444a6a5c534674b4642a1a949799e1d)



> In the latest version, looks like you were missing the Application name entry in the manifest. I added that and when launching you now see the dummy lat/lon values.
> 
> I also went ahead and added the Flows to the wear Activity. This should be very close to a working state in terms of listening to the GnssStatus and Location updates, although it needs some more debugging - I still don't see the status updates on the screen.
> 
> Here's a branch with the changes: https://github.com/barbeau/gpstest/tree/wear-fix-application-manifest
> 
> ...and the commit: [3d440f2](https://github.com/barbeau/gpstest/commit/3d440f234444a6a5c534674b4642a1a949799e1d)

Thank you! It did work. The status doesn't update on the screen because the `gnssStatuses` is empty. It was created by this `val gnssStatuses: List<SatelliteStatus> by signalInfoViewModel.filteredGnssStatuses.observeAsState(emptyList())`. I've studied all the links you shared with me and searched from the internet but still don't know why it failed to observe the `LiveData`.

adaext commented:
> In the latest version, looks like you were missing the Application name entry in the manifest. I added that and when launching you now see the dummy lat/lon values.
> 
> I also went ahead and added the Flows to the wear Activity. This should be very close to a working state in terms of listening to the GnssStatus and Location updates, although it needs some more debugging - I still don't see the status updates on the screen.
> 
> Here's a branch with the changes: https://github.com/barbeau/gpstest/tree/wear-fix-application-manifest
> 
> ...and the commit: [3d440f2](https://github.com/barbeau/gpstest/commit/3d440f234444a6a5c534674b4642a1a949799e1d)
I guess it may be because the `gnssStatuses` list should be initialized by real GNSS data which the simulator couldn't provide. Could you please help me to test the latest commit on a real device? I don't have an electronic watch.

barbeau commented:
>I guess it may be because the gnssStatuses list should be initialized by real GNSS data which the simulator couldn't provide. Could you please help me to test the latest commit on a real device? I don't have an electronic watch.

@adaext I took a closer look, and there is actually an exception being thrown by the `SharedLocationManager`. I added a Log statement in the try/catch block:

```
        try {
            locationManager.requestLocationUpdates(
                LocationManager.GPS_PROVIDER,
                minTimeMillis(context, prefs),
                minDistance(context, prefs),
                callback,
                context.mainLooper
            )
        } catch (e: Exception) {
            Log.e(TAG, "Exception in location flow: $e")
            close(e) // in case of exception, close the Flow
        }
```

...and here's what is says:

>Exception in location flow: java.lang.SecurityException: uid 10080 does not have android.permission.ACCESS_COARSE_LOCATION or android.permission.ACCESS_FINE_LOCATION.

So you just need to ask the user for location permissions prior to calling `gpsStart()` in the `MainActivity`. You should be able to reproduce this on the Wear emulator and still get location updates there. See the app `MainActivity` for code to ask for permissions, although you'll need to double-check that this is the same on Wear.

Also, could you add this type of Log statements in any existing try/catch block in all the Shared*Manager classes so we can easily see when a problem happens?

adaext commented:
> 

@barbeau Thanks! After adding location permissions requests from users, the gnssStatuses list did appear. But they gathered in the same col on each row. Since I've used `scalingLazyColumn` to display them, I don't know how to separate them on each item.

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,431 @@
+        } else {
+            LibUIUtils.showLocationPermissionDialog(this)
+        }
+        gpsStart()
```
`gpsStart()` should only be called after permissions are granted (see the workflow in main app). So you need to check if permissions are granted here - if yes, call `gpsStart()`, if not, call `requestPermissions()`, and then in `onRequestPermissionsResult()` if permissions are granted call `gpsStart()`.

adaext replied:
> 

Ok, thanks!

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,431 @@
+@OptIn(ExperimentalCoroutinesApi::class)
+@Composable
+fun WearApp(satStatues: List<String>, signalInfoViewModel: SignalInfoViewModel) {
+    val gnssStatuses: List<SatelliteStatus> by signalInfoViewModel.filteredGnssStatuses.observeAsState(emptyList())
```
To get `Location` updates to populate the lat/lon/alt/etc., you'll need another observer here, like:

```kotlin
val location: Location by signalInfoViewModel.location.observeAsState(Location("default"))
```

Then every time the `location` in the ViewModel changes you'll get an update here.

adaext replied:
> To get `Location` updates to populate the lat/lon/alt/etc., you'll need another observer here, like:
> 
> ```kotlin
> val location: Location by signalInfoViewModel.location.observeAsState(Location("default"))
> ```
> 
> Then every time the `location` in the ViewModel changes you'll get an update here.

Ok, thanks!

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,431 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        progressBar = findViewById(R.id.progress_horizontal)
```
This `progressBar` doesn't do anything because there is no XML layout and you should get an NPE when referencing it.

Instead of doing this, you should create a composable within `WearApp` that is a progress indicator (does Wear have a native one?).

And, you'll need to create another observer in WearApp for getting notified when fix is lost of obtained, like this:

```kotlin
val fixState: FixState by signalInfoViewModel.fixState.observeAsState(FixState.NotAcquired)
```

adaext replied:
Thanks!

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,419 @@
+                if (!listState.isScrollInProgress) {
+                    TimeText(
+                        timeSource = TimeTextDefaults.timeSource(
+                            DateFormat.getBestDateTimePattern(Locale.getDefault(), "hh:mm:ss")
```
This timestamp should be set based on the `location.getTime()` value.

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,419 @@
+@Composable
+fun WearApp(satStatues: List<String>, signalInfoViewModel: SignalInfoViewModel) {
+    val gnssStatuses: List<SatelliteStatus> by signalInfoViewModel.filteredGnssStatuses.observeAsState(emptyList())
+    val location: Location by signalInfoViewModel.location.observeAsState(Location("default"))
```
`location` is currently unused here - you'll need to update the Composables with data from this object to show `Location` data on the screen.

adaext replied:
> `location` is currently unused here - you'll need to update the Composables with data from this object to show `Location` data on the screen.

Thanks, I'll.

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,419 @@
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@Composable
+fun WearApp(satStatues: List<String>, signalInfoViewModel: SignalInfoViewModel) {
```
You can get rid of `satStatues` here

adaext replied:
> You can get rid of `satStatues` here

Ok, thanks.

barbeau commented on wear/src/main/java/com/android/gpstest/wear/MainActivity.kt:
```diff
@@ -0,0 +1,419 @@
+                    CustomLinearProgressBar(fixState)
+                }
+
+                for (satStatue in satStatues) {
```
This FOR block needs to be replaced with data from the `location` object, which will be updated on each new GNSS fix.

adaext replied:
> This FOR block needs to be replaced with data from the `location` object, which will be updated on each new GNSS fix.

Done
